miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultFileAttachmentService.java

```diff
@@ -0,0 +1,198 @@
+
+  @Value("${miso.fileStorageDirectory}")
+  private String fileStorageDirectory;
+
+  private final Map<String, Function<Long, Attachable>> entityFetchers = new HashMap<>();
+
+  public DefaultFileAttachmentService() {
+    entityFetchers.put("pool", WhineyFunction.rethrow(id -> poolService.get(id)));
+    entityFetchers.put("project", WhineyFunction.rethrow(id -> projectService.getProjectById(id)));
+    entityFetchers.put("run", WhineyFunction.rethrow(id -> runService.get(id)));
+    entityFetchers.put("servicerecord", WhineyFunction.rethrow(id -> serviceRecordService.get(id)));
+  }
+
+  public void setAttachableStore(AttachableStore attachableStore) {
+    this.attachableStore = attachableStore;
+  }
+
+  public void setProjectService(ProjectService projectService) {
+    this.projectService = projectService;
+  }
+
+  public void setRunService(RunService runService) {
+    this.runService = runService;
+  }
+
+  public void setServiceRecordService(ServiceRecordService serviceRecordService) {
+    this.serviceRecordService = serviceRecordService;
+  }
+
+  public void setAuthorizationManager(AuthorizationManager authorizationManager) {
+    this.authorizationManager = authorizationManager;
+  }
+
+  public void setFileStorageDirectory(String fileStorageDirectory) {
+    this.fileStorageDirectory = fileStorageDirectory;
+  }
+
+  @Override
+  public Attachable get(String entityType, long entityId) {
+    Function<Long, Attachable> fetcher = entityFetchers.get(entityType);
+    if (fetcher == null) {
+      throw new IllegalArgumentException("Unknown entity type: " + entityType);
+    }
+    return fetcher.apply(entityId);
+  }
+
+  @Override
+  public void add(Attachable object, MultipartFile file) throws IOException {
+    Attachable managed = attachableStore.getManaged(object);
+    String saveFilename = Long.toString(new Date().getTime());
+    String relativeDir = makeRelativeDir(object.getAttachmentsTarget(), object.getId());
+    File saveDir = new File(makeFullPath(relativeDir));
+    File targetFile = new File(saveDir, saveFilename);
+    while (targetFile.exists()) {
+      saveFilename = Long.toString(Long.valueOf(saveFilename) + 1);
+      targetFile = new File(saveDir, saveFilename);
+    }
+    if (LimsUtils.checkDirectory(saveDir, true)) {
+      file.transferTo(targetFile);
+    } else {
+      throw new IOException("Cannot upload file - check that the directory specified in miso.properties exists and is writable");
+    }
+
+    try {
+      FileAttachment attachment = new FileAttachment();
+      attachment.setFilename(file.getOriginalFilename());
+      attachment.setPath(relativeDir + File.separator + saveFilename);
+      attachment.setCreator(authorizationManager.getCurrentUser());
+      attachment.setCreationTime(new Date());
+      managed.getAttachments().add(attachment);
+      attachableStore.save(managed);
+    } catch (Exception e) {
+      if (!targetFile.delete()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Use "java.nio.Files#delete" here for better messages on error conditions. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4042)